Add sample bank selection controls (and hotkeys) to editor

peppy opened the pull request:
Mostly for the hotkeys, since they've come up multiple times.


https://user-images.githubusercontent.com/191335/196695476-fa20b4e7-c646-4114-aaa9-3aadf3f2e82c.mp4

- [x] Hotkeys need to be fixed to work post-merge of #20903

peppy committed:
Add and consume sample bank constants

peppy committed:
Add sample bank suport to editor selection handler

peppy committed:
Add icon and radio button logic

peppy committed:
Ensure placement uses currently selected bank

peppy committed:
Fix bank potentially being overwritten during placement

peppy committed:
Fix huge oversight causing test failures

bdach commented on osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs:
```diff
@@ -104,12 +164,48 @@ protected virtual void UpdateTernaryStates()
+        /// Removes a sample bank from all selected <see cref="HitObject"/>s.
+        /// </summary>
+        /// <param name="bankName">The name of the sample bank.</param>
+        public void RemoveSampleBank(string bankName)
```
Method is not actually used. Not sure it makes any sense either?

peppy replied:
I've removed this.

bdach commented on osu.Game/Screens/Edit/Compose/Components/EditorSelectionHandler.cs:
```diff
@@ -48,11 +49,70 @@ private void load()
+                    Description = bankName.Titleize()
+                };
+
+                bindable.ValueChanged += state =>
```
Personal preference: This `ValueChanged` callback is a bit on the longer side. I'd probably split a method.

peppy replied:
Hmm, I dunno about this. I don't think splitting out is going to do much for readability given what the method is doing..

bdach commented on osu.Game/Screens/Edit/Compose/Components/ComposeBlueprintContainer.cs:
```diff
@@ -186,6 +207,39 @@ protected virtual IEnumerable<TernaryButton> CreateTernaryButtons()
+                        X = 10,
+                        Y = -1,
+                        Font = OsuFont.Default.With(weight: FontWeight.Bold, size: 20),
+                        Text = $"{char.ToUpper(sampleName.First())}"
```
```suggestion
                        Text = $"{char.ToUpperInvariant(sampleName.First())}"
```

A little safer. Guess I didn't ban this one back when...

peppy commented:
> Just a quick feedback, but it definely should be possible to select different banks for addtitions (whistle, finish, clap) and the hitobject itself, as combining those is a huge part of the hitsound stack and most of the hitsounds nowdays use it.

That has nothing to do with this change, right? People were asking for shortcut level bank switching and this adds that.

peppy commented on osu.Game/Rulesets/Edit/HitObjectComposer.cs:
```diff
@@ -146,6 +148,16 @@ private void load(OverlayColourProvider colourProvider)
                                         Spacing = new Vector2(0, 5),
                                     },
+                                },
+                                new EditorToolboxGroup("bank (Shift-Q~R)")
```
Brings up a good point - this is misaligned due to no "auto".

peppy replied:
I've added back automatic selection and now things are realigned.

peppy committed:
Use `ToUpperInvariant` for added safety

Co-authored-by: Bartłomiej Dach <dach.bartlomiej@gmail.com>

peppy commented:
Blocking this temporarily while I figure out what to do about "auto" (which doesn't really exist in lazer).

peppy commented:
Note to self: `RadioButton` class exists and may be a better fit for this.

peppy committed:
Merge branch 'master' into sample-set-controls

peppy committed:
Update new usages of "soft" to use the new constant

peppy committed:
Update to support non-control-point sample changes

peppy committed:
Fix hotkeys not actually working

peppy committed:
Don't use `switch` for single `case` statement

peppy committed:
Add "auto" bank selection during placement

peppy committed:
Add test coverage of new hotkeys

peppy commented:
I've tidied this up (as much as it can be) and added test coverage.

I foresee this logic changing a bit once https://github.com/ppy/osu/pull/23445 is in, but I see both systems working together (just being more specific about where the banks are assigned to based on context). Failing that, we might be able to pull out this UI and use it in the popover as well (almost did that in this PR but feel we should get a base implementation in first).

peppy commented:
The fail case here is where a `HitObject` has no samples, and the `All` condition here becomes `true` as a result:

![JetBrains Rider 2023-05-25 at 08 51 40](https://github.com/ppy/osu/assets/191335/82099f09-eca2-4829-be6d-7fbc33c525bc)

This can only happen in tests, but I'll look to fix the logic regardless.

peppy committed:
Fix bindable feedback loop

bdach committed:
Change `PlacementBlueprint.AutomaticBankAssignment` to property

Mostly for consistency.

bdach committed:
Fix `ComposeBlueprintContainer` briefly assigning `auto` bank

Seems to have had no consequence due to the way
`AutomaticBankAssignment` works (that flag is checked in
`PlacementBlueprint.UpdateTimeAndPosition()`, which runs essentially
every frame), but let's avoid putting it there at all ever.

bdach committed:
Merge branch 'master' into sample-set-controls

bdach commented:
I've pushed in a few menial changes. Of particular note, https://github.com/ppy/osu/pull/20829/commits/7d8f08c0ea0e2e1f6d2952926f85557ff9447350 fixed a thankfully-non-issue which had `ComposeBlueprintContainer` briefly apply the `auto` bank to hitobjects. This was remedied because the automatic bank assignment feature runs essentially in `Update()`, so that `auto` bank was never user-facing, but I figure let's just avoid putting it there at all ever if it can be avoided.

There are a few wonky spots here, in particular I'm not sure how the automatic assignment feature is supposed to work with volume. The way it kinda works right now is that using automatic assignment will inherit not only the bank, but the volume as well, and if you switch to another bank then that volume will be kept - but if you, for instance:

1. choose auto bank
2. switch to soft
3. place an object

the placed object will inherit the volume of the previous object from the auto bank choice, but the new object pending placement after it will have 100 volume, which is probably not what is wanted. However, I'm not sure *what* is, and this feels like an important enough feature wherein I'm willing to look away, get this in, and see what mappers' expectations and feedback is going to be based on the way they'll interact with this. Or you could open another pull on top of this fixing that issue if you think you know a way to solve it. Either works for me.
